packages/flutter/lib/src/widgets/conduit.dart

```diff
@@ -0,0 +1,375 @@
+// mountPoint, chances are the [RenderConduitAnchorWidget] hasn't been built for
+// the first time yet. In this case a _PendingRenderObjectManifest will be given
+// to the [RenderConduitAnchorWidget] that has yet to build, allowing it to attach
+// the pending renderObject when it's ready.
+@immutable
+class _PendingRenderObjectManifest {
+  const _PendingRenderObjectManifest(this.renderObject, this.leaderRenderBox)
+    : assert(renderObject != null),
+      assert(leaderRenderBox != null);
+
+  final RenderObject renderObject;
+  final _RenderLeaderProxyBox leaderRenderBox;
+}
+
+class _BoxContainer<T> {
+  T content;
+}
+
+/// A global key that uniquely identifies the receiving [RenderConduitAnchorWidget]
+/// that a [Conduit] will later attach its `remoteChild`'s `renderObject` to.
+///
+/// Unlike other global keys, an [AnchorKey] may contain opaque information only
+/// useful to [RenderConduitAnchorWidget]s, so it typically shouldn't be used to
+/// key other types of widgets.
+class AnchorKey extends GlobalKey {
+  /// Creates a unique [AnchorKey].
+  AnchorKey()
+    : _manifest = _BoxContainer<_PendingRenderObjectManifest>(),
+      super.constructor();
+  final _BoxContainer<_PendingRenderObjectManifest> _manifest;
+
+  @override
+  _RenderConduitAnchorElement get currentContext => super.currentContext as _RenderConduitAnchorElement;
+}
+
+/// A widget that builds its [child] and [remoteChild] as regular children under
+/// itself, but places its [remoteChild]'s render object to the [RenderConduitAnchorWidget]
+/// associated with [remoteKey].
+///
+/// This widget is useful when you need to build the [remoteChild] subtree using
+/// an [OverlayEntry], but would like the [remoteChild] to depend on the same set
+/// of [InheritedWidget]s as this widget. For example, an in flight [Hero] needs
+/// to be placed on an [OverlayEntry] so it appears over other content, but at
+/// the same time we want the [DefaultTextStyle] it depends on to remain the same
+/// when it starts flying.
+class Conduit extends RenderObjectWidget {
+  /// Creates a [Conduit] that places its [remoteChild]'s render object to the
+  /// designated [RenderConduitAnchorWidget] .
+  const Conduit({
+    Key key,
+    AnchorKey remoteKey,
+    this.remoteChild,
+    this.child,
+  }) : remoteKey = remoteChild == null ? null : remoteKey,
+       // A remoteChild must be accompanied by a remoteKey.
+       assert(remoteChild == null || remoteKey != null),
+       super(key: key);
+
+  /// A global key that uniquely identifies a [RenderConduitAnchorWidget].
+  ///
+  /// Must not be null when [remoteChild] is not null.
+  ///
+  /// Currently, directly changing [remoteKey] from a non-null value to a different
+  /// non-null value (i.e. reparenting [remoteChild]) is not supported. To switch
+  /// to a different [remoteKey], first set this property to null and then change
+  /// it to the new [remoteKey].
```

Comment:
What's the difficulty in supporting key changes?

Reply:
For my use case I don't think it's needed (the remote child should be reusing the same `OverlayEntry`) so I didn't bother implementing that. I think it can be added as a new feature when needed?